external/model-preparation-algorithm/mpa_tasks/utils/runner.py

```diff
@@ -41,6 +40,38 @@ def stop(self) -> bool:
             self._max_epochs = self.epoch
         return broadcast_obj[0]
 
+    @check_input_parameters_type()
+    def train(self, data_loader: DataLoader, **kwargs):
+        self.model.train()
+        self.mode = "train"
+        self.data_loader = data_loader
+        self._max_iters = self._max_epochs * len(self.data_loader)
+        self.call_hook("before_train_epoch")
+        # TODO: uncomment below line or resolve root cause of deadlock issue if multi-GPUs need to be supported.
+        # time.sleep(2)  # Prevent possible multi-gpu deadlock during epoch transition
+        for i, data_batch in enumerate(self.data_loader):
+            self._inner_iter = i
+            self.call_hook("before_train_iter")
+            self.run_iter(data_batch, train_mode=True, **kwargs)
+            self.call_hook("after_train_iter")
+            if self.stop():
+                break
+            self._iter += 1
+        self.call_hook("after_train_epoch")
+        self.stop()
+        self._epoch += 1
+
+
+@RUNNERS.register_module()
+class ClsEpochRunnerWithCancel(EpochRunnerWithCancel):
+    """
+    Classification runner based on EpochRunnerWithCancel
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.save_ema_model = False
```

Comment:
Is overriding new runner class the best solution on our scenario?
Could you explain us why runner should have this attribute, not on hook or stage or trainer?

Reply:
Because it is the attribute of the runner class and it is right and good practice to declare an attribute in the related class, not on the go somewhere

Reply:
Hmm, I can't understand well. Is there no solution without modifying runner?
You said current runner implementation that uses exactly same code on each task is not reasonable because they are separated. I agree with you. However, if then, I think we should avoid updating runner as possible to gather runners in one file. How do you think?

Reply:
For my opinion it is a bad practice to create attributes on the go somewhere in the code without a declaration in the base class. But, we can do it in eval hook, of course. https://github.com/openvinotoolkit/model_preparation_algorithm/pull/73/files

Please, be aware that now we don't use this runner for task. The mmdetetion object is created from registry and is utilized in the task.
To avoid this, please, consider to delete duplicated runners in mmdet and mmseg

Reply:
I prefer to have any classes clearly declared. Getting the `AttributeError` somewhere is bad enough. Not seeing clearly the lifecycle of an instance attribute is worse. In some cases a lot of time required to restore possible call sequences. Code is read more often than written. So, I would perfer to avoid such cases. 
We have a lot things in MPA depending on each other and impacting other tasks. Let's start to make code more undependable.

Reply:
Thank you for your kind explanation and effort :) I think I understand all of things. I'll resolve this conversation.